### DISCUSSION TITLE
HttpError: Not found

### DISCUSSION THREAD

**Discussion by henild** [octokit/octokit.js#1974] [Q&A] [answered]
I am trying to do something like this:
```ts
    const response = await octokit.request('GET /user/starred/{owner}/{repo}', {
        owner: "octocat",
        repo: "hello-world"
    })
```

but getting error
```log
error: Uncaught (in promise) HttpError: Not Found
        const error = new RequestError(message, status, {
                      ^
    at https://cdn.skypack.dev/-/@octokit/request@v5.4.12-7VbU50b4qMn4Kz3Eh65V/dist=es2020/@octokit/request.js:65:23
    at async test.ts:18:22
```

Simple endpoints like `/user/starred`, `/user` are working fine. I tried with `curl` but it's not working there too. Is this a problem with github API?

**Comment by gr2m** [comment]
I was able to reproduce the problem: https://runkit.com/gr2m/octokit-rest-js-1974/2.0.0

I feel like I must be missing something, but it really looks like an API bug. Hard to believe that it's something that wouldn't be caught by the API team. I'll reach out through internal channels and keep you posted, thank you for reporting

**Comment by gr2m** [comment, marked as answer]
Turns out it's expected behavior, but I agree it's confusing.

See [the docs](https://docs.github.com/en/free-pro-team@latest/rest/reference/activity#check-if-a-repository-is-starred-by-the-authenticated-user)

<img width="762" alt="image" src="https://user-images.githubusercontent.com/39992/102721053-273cbf00-42ad-11eb-8031-4c1e611c8bcd.png">

You can do the following

```js
const isStarred = await octokit.request('GET /user/starred/{owner}/{repo}', {
    owner: "octocat",
    repo: "hello-world"
}).then(() => true, error => {
  if (error.status === 404) return false
  throw error
})
```

**Reply by henild** [reply]
Thanks, now it doesn't give an error. But I am trying to reach this endpoint `PUT /user/starred/{owner}/{repo}` it is always returning false. i.e. Resource Not found even for valid repositories. Am I missing something or the endpoint is broken?

**Reply by gr2m** [reply]
The `.then()` handler I shared above only makes senses for `GET /user/starred/{owner}/{repo}`, it turns the different responses into `true`/`false` values. Don't use it for `PUT /user/starred/{owner}/{repo}`. If it returns a 404, then the repository doesn't exist, or it's a private repository and you don't have access to it with the provided authentication token

**Reply by henild** [reply]
Actually, I checked thrice, I am sure the authentication token is valid and the repo is public.

If I don't use `.then()`, then it's again giving the previous error. But I think that should also work fine(at least that's what docs say) here, because if it returns `404` then it returns `false` else throw an error (which it is not doing).

Sorry if this is not a proper place for this, but I don't have any better place to ask this question. If you know do let me know.

EDIT: Just to be clear, it's giving `Not found` even with the example giving in the docs (with both curl and octokit)

```
curl \
  -X PUT \
  -H "Accept: application/vnd.github.v3+json" \
  -u username:key \
  https://api.github.com/user/starred/octocat/hello-world
  ```

**Reply by gr2m** [reply]
instead of `-u username:key`, can you please try `-H "Authorization: token [key]"`

**Reply by henild** [reply]
Same result with that too.

**Reply by gr2m** [reply]
this request works for me:

```
curl \
  -X PUT \
  -H "Accept: application/vnd.github.v3+json" \
  -H "Authorization: token $GITHUB_TOKEN" \
  https://api.github.com/user/starred/octocat/hello-world
```

**Reply by henild** [reply]
Sigh... Turns out it was a permission issue. I had given only permission to `Update all user data`. And didn't think the public repository require to enable that permission manually(this is really confusing). 

The top section of the permission says `Full control of private repositories ` but it contains `Access public repositories`. If you can please forward this to the API team about the confusion.

**Reply by gr2m** [reply]
The response headers provide a clue. Not always, but for many API endpoints you see these two headers:

```
X-OAuth-Scopes: user
X-Accepted-OAuth-Scopes: public_repo, repo
```

The `X-OAuth-Scopes` response headers tells you which scopes your token has, `X-Accepted-OAuth-Scopes` tells you a list of scopes, one of which is required depending on the use case.

You are not the first to be confused by the `404` errors. Maybe we should log a more helpful message pointing out that the scopes are not sufficient for the request.
